### ISSUE TITLE
Improve ipfs add performance for large trees of files & directories

### ISSUE THREAD

**Issue by dirkmc** [ipfs/go-ipfs#6523]
Calling `ipfs add` on a directory with a large tree of sub-directories and files is slow. This use case is [particularly important for file-system based package managers](https://github.com/ipfs/package-managers/issues/18).

#### Background ####

IPFS deals with immutable blocks. Blocks are stored in the blockstore.
The [UnixFS package](https://github.com/ipfs/go-unixfs) breaks files up into chunks, and converts them to IPLD objects.
The [DAG Service](https://github.com/ipfs/go-ipld-format) stores IPLD objects in the blockstore.

The [Mutable File System (MFS)](https://github.com/ipfs/go-mfs) is an abstraction that presents IPFS as a file system. For example consider the following directory structure with associated hashes:

```
animals         <Qm1234...>
  land          <Qm5678...>
    dogs.txt    <Qm9012...>
    cats.txt    <Qm3456...>
  ocean         <Qm7890...>
    fish.txt    <Qm4321...>
```

If the contents of `fish.txt` changes, the CID for `fish.txt` will also change. The link from `ocean → fish` will change, so the CID for `ocean` will change. The link from `animals → ocean` will change so the CID for `animals` will change. MFS manages those links and the propagation of changes up the tree.

#### Algorithm ####

`ipfs add` uses the MFS package to add files and directories to the IPFS blockstore. To add a directory with a large tree of sub-directories and files:
- Create an MFS root for the root directory (`animals` in the example above)
- Recurse through the directory structure in "depth first search" fashion.
  For each directory
  - Create a corresponding empty directory in MFS eg `animals/ocean`
    This [adds the empty directory](https://github.com/ipfs/go-mfs/blob/master/dir.go#L311) to the blockstore.
  - For each file in the directory eg `animals/ocean/fish.txt`
    - [Read](https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L379) the file contents
    - [Convert the contents](https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L101) into a chunked IPLD node
    - [Add the IPLD Node](https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L129) and all its chunks to the blockstore
    - Create the directory in MFS, if it doesn't exist (†) eg `animals/ocean`
    - Add the IPLD Node representing the file to MFS at the correct path (eg `animals/ocean/fish.txt`)
      _Note: This again [adds the IPLD Node](https://github.com/ipfs/go-mfs/blob/master/dir.go#L358) root to the blockstore_
- Recurse through the MFS representation of the directory structure
  - For each directory, call `directory.GetNode()`
    Note that at this stage, the links to files in the directories have been created, so the directory created here will have a different CID than the empty directory created *before* the files were added. Calling `directory.GetNode()` (confusingly) *[writes the directory with links to files](https://github.com/ipfs/go-mfs/blob/master/dir.go#L444)* to the blockstore

(†) *Although we've already created the directory, it's necessary to again ensure it exists before adding the file, because after processing every 256k files, the MFS internal directory cache structure is dereferenced to allow for golang garbage collection*

#### Areas for Improvement ####

- The IPLD Node root for a file is added to the blockstore twice
  - When the file is converted to an IPLD node
  - When the file is added to MFS
- The MFS directory structure is kept in memory
- Recursion over the directory structure happens twice
  - While reading the structure from input
  - While writing out the directories to the blockstore
- It appears that in some circumstances branches of the directory structure will not be written to the blockstore
  - After processing every 256k files, the MFS internal directory cache is cleared
  - Each time a file is added, parent MFS directories are created if they don't exist
  - This means that branches of the directory tree whose files have all been written will not be recreated after clearing the directory cache
  - The second recursion over the directory structure will therefore miss these branches and they will not be written to the blockstore
- The progress indicator pauses for a long period when it is almost at 100% while the directories are being written to the blockstore

#### Proposed Improvements ####

The above issues would be solved if we interact directly with the UnixFS API instead of with the MFS API:
- Recurse once over the directory structure
- Add files as we go
- Add directories once all their files have been added

#### Future Work ####

It has been noted that [disk throughput](https://github.com/ipfs/package-managers/issues/18#issuecomment-471125150) and [CPU usage](https://github.com/ipfs/package-managers/issues/18#issuecomment-471141763) are not close to maximum while adding large numbers of files. Future work should focus on analyzing these findings.

**Comment by dirkmc** [comment]
@Stebalien & @magik6k this analysis came from spelunking through the source code. Please let me know if there's anything there that doesn't sound right. In particular I'd like a second opinion on the theory (untested) that branches of the directory tree may be missed when writing the directories to the blockstore

**Comment by Stebalien** [comment]
This is mostly correct.

> The IPLD Node root for a file is added to the blockstore twice 

This should be slightly mitigated by the fact that we _check_ if we already have a block before writing it (in the blockservice itself).

> The second recursion over the directory structure will therefore miss these branches and they will not be written to the blockstore

Not quite. We _flush_ everything first. Unfortunately, this _does_ mean that we end up storing a bunch of intermediate directories that we don't need.

---

For some context, we _primarily_ use MFS to make it easy to pause for GC. However, honestly, that probably isn't worth it given the performance hit. I agree we should explore using unixfs directly.

**Comment by dirkmc** [comment]
> We flush everything first

Do you mean when we [call `root.FlushMemFree()`](https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L340)? It [calls `dir.Flush()`](https://github.com/ipfs/go-mfs/blob/master/root.go#L170) but AFAICT that will only [Flush the current dir](https://github.com/ipfs/go-mfs/blob/master/dir.go#L339) and all dirs up the tree to the root (not any previously added branches of dirs)

**Comment by Stebalien** [comment]
It will recursively flush. Flush calls [`dir.GetNode()`](https://github.com/ipfs/go-mfs/blob/master/dir.go#L340) which calls [`dir.sync()`](https://github.com/ipfs/go-mfs/blob/master/dir.go#L434) which walks all the cached entries, calling GetNode on each.

Finally, it calls `GetNode()` on the unixfs directory itself then adds that to the DAGService.

**Comment by Stebalien** [comment]
(unless I'm missing something)

**Comment by dirkmc** [comment]
Ah yes you're right, thanks 👍

I'll remove that section from the original issue above.

**Comment by dirkmc** [comment]
@Stebalien given this analysis, my uneducated guess is that changing `ipfs add` to use the UnixFS API directly is going to give us a moderate performance improvement, maybe on the order of 10% - 20%. Would your guess fall somewhere in that range?

**Comment by magik6k** [comment]
We should measure how much datastore is called when adding stuff with different layouts.

* Count Put/Get/Has when adding
  * Many tiny files (ideally above the 256k treshold, but that may not be practical)
  * 1000 100kb files
    * in lots of directoires
  * 100 1mb files
  * 10 10mb files
  * 1 100mb file
  * (some variations based on directory depth)

My bet would be that for large directories / lots of directories, `Has` calls will create a significant overhead.

**Comment by magik6k** [comment]
Another note on using unixfs directly - It would make it easier to add small files in parallel - each call to `addFile` could try to buffer say 100kb of data, and if an entire file fits in that amount of data, add it async (which, when adding lots of small files should help with waiting on IO)

**Comment by dirkmc** [comment]
I was thinking along those lines - maybe a queue for incoming files and a queue for outgoing with backpressure in each

**Comment by dirkmc** [comment]
> we primarily use MFS to make it easy to pause for GC

@Stebalien can we just add the current root for any in-progress `ipfs add` to [bestEffortRoots](https://github.com/ipfs/go-ipfs/blob/master/pin/gc/gc.go#L41)?

**Comment by Stebalien** [comment]
Probably? We really do need some kind of internal in-memory pinning system (which shouldn't be too hard). The primary issue is that the importer would need to be GC aware so it can pause.

**Comment by dirkmc** [comment]
We can just do something similar to how `ipfs add` [works now](https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L428), right, ie check periodically if GC has been requested, and if so, create a temporary fake tree and pin the root.

**Comment by Stebalien** [comment]
Yeah, you're right.

**Comment by Kubuxu** [comment]
If there is a significant `Has` performance hit, the bloom filter cache in  blockstore will help a lot.

**Comment by magik6k** [comment]
In most cases where we call `Has` while adding, the thing is already in the blockstore (likely in all cases), so Bloom filter won't really help here since we still need to check in the underlying blockstore. 

(hack / poc idea: check if `has` always returns True when adding, if that's the case, create blockstore wrapper which replaces `Has` with hardcoded `return true`. If that works we can quickly measure the overhead of all the `has` calls, and see if we're fixing hhe right thing)

**Comment by Kubuxu** [comment]
`Has` is called before adding a block to blockstore (always). If you always return true it won't add the files to datastore.

**Comment by magik6k** [comment]
Yeah, that happens in blockservice, and it's easy to disable that check there (or just override `Has` in blockservice)

(https://github.com/ipfs/go-blockservice/blob/master/blockservice.go#L141, https://github.com/ipfs/go-blockservice/blob/master/blockservice.go#L173)

**Comment by dirkmc** [comment]
These data were generated with this [script](https://github.com/dirkmc/go-ipfs/blob/exp/add-perf/core/coreunix/addperf_test.go) running on my laptop. It creates `count x <file size>` files of random data in a single directory (in memory), creates an Adder with an in-memory datastore and calls `AddAllAndPin()` on the directory.

```
       Name          Has      Get      Put   Objects
 1 x      100M:     1230        3      411         2    281.674222ms
 10 x      10M:     1248       12      417        11    250.425109ms
 100 x      1M:     1518      102      507       101    260.626669ms
 1024 x   100k:     3090     1026     1031      1025    375.101811ms
 10240 x   10k:    30738    10242    10247     10241    1.314768727s
 20480 x    5k:    61458    20482    20487     20481    3.366398431s
 51200 x    2k:   153618    51202    51207     51201    27.019718374s
 68266 x  1.5k:   204816    68268    68273     68267    1m2.91561757s
 81920 x 1.25k:   245778    81922    81927     81921    2m5.529349576s
 102400 x   1k:   307218   102402   102407    102401    4m6.188296442s
```

`Objects` is the number of Objects output to the response stream.

**Comment by Kubuxu** [comment]
We really need tracing in those code paths, this would let us know what is the precise cause.

It might be worth comparing that with badger.

**Comment by dirkmc** [comment]
@Kubuxu good call 👍

I'm continuing to work on generating some useful profiling information, I just wanted to get this out here as a start as @magik6k had some questions on the how many calls we're making to Has/Get/Put

**Comment by dirkmc** [comment]
Also please let me know if there's other particular things you'd like to see stats about, you guys are more familiar with where bottlenecks most likely to be

**Comment by Kubuxu** [comment]
You might also want to get stats from `localhost:5001/debug/metrics/prometheus` it will have stats on how much time total was spent in datastore in a given call.

**Comment by dirkmc** [comment]
I added the timings for Badger DB to the table in [my comment above](https://github.com/ipfs/go-ipfs/issues/6523#issuecomment-512951980)

**Comment by Kubuxu** [comment]
Very interesting.

**Comment by dirkmc** [comment]
I wanted to tested the impact of adding many files in a directory tree with a varying branch out factor
For example 27 files in a directory tree with a branch out factor of 3 would have 3 levels of directories, with all the files at the leaves.

It doesn't seem to have a big performance impact.

In these graphs the branch out factor (items / directory) is on the x-axis and duration is on the y-axis, where each line represents a `<total number of files> x <file size>`:

<img width="605" alt="Screen Shot 2019-07-19 at 1 52 29 PM" src="https://user-images.githubusercontent.com/169124/61555912-720a3580-aa2e-11e9-8b9d-85bf545a83cf.png">

These graphs show Has / Get / Put / Object counts for 68,266 1.5k files and for 102,400 1k files.

<img width="604" alt="Screen Shot 2019-07-19 at 2 17 02 PM" src="https://user-images.githubusercontent.com/169124/61556492-eabdc180-aa2f-11e9-9603-1743d2465e5a.png">


([Spreadsheet](https://docs.google.com/spreadsheets/d/15UU5lxUt6dwap29o5LBIYKexPBhM9bTTjCqIPzt6Sf4/edit?usp=sharing) / [Code](https://github.com/dirkmc/go-ipfs/blob/9977e14119440cde116a9de7d271dded3d570a97/core/coreunix/addperf_test.go))

**Comment by dirkmc** [comment]
I ran some simulations for varying numbers of files with a few different branch-out factors. Again branch-out factor doesn't appear to make a difference, and it appears that time complexity increases polynomially ([Spreadsheet](https://docs.google.com/spreadsheets/d/15UU5lxUt6dwap29o5LBIYKexPBhM9bTTjCqIPzt6Sf4/edit?usp=sharing) - See Sheet 2).

<img width="605" alt="Screen Shot 2019-07-21 at 11 26 27 AM" src="https://user-images.githubusercontent.com/169124/61593201-6f344f80-abaa-11e9-955a-e2a5076d0faf.png">

**Comment by warpfork** [comment]
Are there any good comparisons we could make with things that are non-ipfs, to get some other baselines to contrast things with?

I'm loving these graphs, and they're super awesome for helping us see some things about our own code paths, but after ogling for a while, I realized the number of zeros on most of the axes are hard for my brain to parse -- are these comparable to how fast a plain filesystem would flush?  If there's a constant factor difference, that's useful to know; the rough order of mag of that constant factor would also be useful to know.

(plain filesystem flush might not be a fair comparison -- we're hashing, etc, And That's Different; maybe 'git add' or operations like that would be better?  There's also a pure golang git implementation out there, which could be amusing if we want to avoid comparing directly to one of the most hyper-optimized c things on the globe.)

**Comment by dirkmc** [comment]
That's a good question - I like the idea of comparing with git, and actually I think the fastest implementation is probably a good baseline.

With these graphs I'm mostly trying to figure out where the performance bottlenecks are with our implementation, eg
- is there a problem adding a directory tree with a large branch-out factor (seems like no)
- does performance degrade as file count increases (seems like yes)

But I think you're right it would be good to compare against realistic maximum performance of a comparable system

**Comment by Stebalien** [comment]
Can we re-run those tests on badger?

**Comment by postables** [comment]
`rsync` is probably a useful tool for comparison with. Maybe `scp` as well

**Comment by dirkmc** [comment]
@Stebalien good idea - I will re-run the tests for the last graph (File Count vs Duration for files of size 1k) on Badger overnight

**Comment by dirkmc** [comment]
<img width="604" alt="Screen Shot 2019-07-22 at 1 09 37 PM" src="https://user-images.githubusercontent.com/169124/61650581-06b8a180-ac82-11e9-830a-1e3622462536.png">

I re-ran the tests against Badger (red line). It seems like performance is also "gently" polynomial, but with a large number of files it's actually better than in-memory storage (blue line), perhaps because there is some overhead keeping that much data in an in-memory map in go?

**Comment by aschmahmann** [comment]
@dirkmc this is looking great. Would you be able to also see how long it takes to do an `ipfs add` using the Filestore?

**Comment by dirkmc** [comment]
As a Proof of Concept I replaced the use of MFS with [direct calls to UnixFS](https://github.com/dirkmc/go-ipfs/commit/5a6227667c37f7b0ad468eb2898dcdd47ebb7b12) (but ignored GC for now). I ran a test against Badger with a branch-out factor of 32, it appears to perform significantly better than the Badger Average graphed above ([Spreadsheet](https://docs.google.com/spreadsheets/d/15UU5lxUt6dwap29o5LBIYKexPBhM9bTTjCqIPzt6Sf4) - See Sheet 2 / [Code](https://github.com/dirkmc/go-ipfs/commit/5a6227667c37f7b0ad468eb2898dcdd47ebb7b12))

<img width="605" alt="Screen Shot 2019-07-22 at 5 41 15 PM" src="https://user-images.githubusercontent.com/169124/61667424-f0254100-aca7-11e9-9bdc-8c0e98e445f2.png">

I'll run tests with different branch-out factors overnight.

**Comment by whyrusleeping** [comment]
By 'use unixfs directly' you mean 'dont create any directories'? I'm willing to believe there's some inefficiency in the mfs code here, but it seems we're comparing apples to oranges, of course it's going to be faster if we just add each file without tracking where it lives

**Comment by whyrusleeping** [comment]
Ah, I see it now. all the commented out stuff threw me for a loop.

Are these tests being run with or without a running daemon?

**Comment by dirkmc** [comment]
So far I'm just hooking directly into [the Adder 🐍](https://github.com/dirkmc/go-ipfs/commit/5a6227667c37f7b0ad468eb2898dcdd47ebb7b12#diff-e11668d0163d94c9fabad547d4e611e2) so as to reduce the amount of noise from other stuff going on in IPFS.

I compared the output hashes of the new code vs the old code and it looks like it's producing the correct output, although the hashes are in a different order as the new code outputs the directory hash as soon as the directory's contents has been added, whereas the old code output all the directories at the end. It's still not quite apples to apples because for the time being I'm ignoring GC.

The real test will be to compare adding a package manager scale directory tree to a running IPFS instance with the old code vs the new code. I'd like to compare that to rsync and git as people suggested above, and also to test with Filestore.

**Comment by whyrusleeping** [comment]
> I compared the output hashes of the new code vs the old code and it looks like it's producing the correct output

ah, perfect. that was my primary concern.

> although the hashes are in a different order as the new code outputs the directory hash as soon as the directory's contents has been added

So, here's a bit of weirdness. one of the main motivators behind using mfs was the fact that the api accepts files in any order. so you could receive `foo/bar/baz` `abc/def/cat.txt` then `foo/bar/bop`. and the current code would handle that fine. Thats just something to keep in mind here, don't let it get in the way of your optimizations and profiling :)

**Comment by Stebalien** [comment]
This is unfortunate. As far as I can tell, js-ipfs will actually do this as you can pass files as a series of `{path: "/foo/bar.txt", content: "..."}`. @alanshaw, is this correct?

**Comment by dirkmc** [comment]
@andrew very helpfully [created a histogram](https://github.com/ipfs/package-managers/issues/79) for a real package manager directory tree, albeit a small one. To make a more realistic simulation I [modified the performance test](https://github.com/dirkmc/go-ipfs/blob/exp/add-perf/core/coreunix/addperf_test.go#L229) to create a directory tree on disk (instead of in-memory) that looks (very roughly) like the histogram:
- 45% of files are 1k in size
- 50% of files are 1k - 1M in size
- 5% of files are 1M - 100M in size
- branch-out factor of 32

I tried profiling the Adder but I found that profiling was too sharp a tool (or at least I couldn't make sense of the output) for a whole sub-system. Currently we read files off the disk and process them one at a time so I decided to try parallelizing file adding.

I tested this for varying numbers of files from 1k - 32k with
- the original Adder
- Adder with direct calls to UnixFS (no MFS)
- Adder with direct calls to UnixFS (no MFS) and parallelization of file adds
  - 8 threads
  - 16 threads
  - 32 threads

The file count / total tree sizes tested were:
```
1k    3G
2k    4.8G
4k    12G
8k    23G
16k   46G
32k   93G
```

My system:
```
$ sysctl hw.physicalcpu hw.logicalcpu
hw.physicalcpu: 4
hw.logicalcpu: 8
```

Performance appears to be roughly linear for this shape directory structure. For my system the optimal number of threads seems to be around 8 - 16, where performance reaches about twice as fast as the current Adder.

<img width="669" alt="Screen Shot 2019-07-25 at 4 41 15 PM" src="https://user-images.githubusercontent.com/169124/61907222-0aa32880-aefb-11e9-844f-aa59c8a7f6b0.png">

([Spreadsheet](https://docs.google.com/spreadsheets/d/1iZFmdK9l__8VMMmVREMMJKpHW0Pue8DJrFRVlos0jJ8/edit?usp=sharing) - [Code](https://github.com/dirkmc/go-ipfs/commit/1c3c4f0a8142bf778500efae4eb0ba1048993f13))

*Please note that as I've tried different experiments the code has become increasingly embarrassing 🙈
The parallelization code is very much proof-of-concept, contains a lot of panics, and probably some races and bugs.*

**Comment by andrew** [comment]
For a real-world benchmark I ran @dirkmc's dirks branch against a copy of the main Arch linux repository and also saw a 2x speed improvement 🚀

## ipfs/go-ipfs master

```
$ time $GOPATH/bin/ipfs add -r --progress --offline --quieter ~/arch
47.50 GiB / 47.50 GiB [========================================================] 100.00%
QmY5C4FXTDtmbyyKX1KixVMedJQ2nk9RgJMv6j9pDzrUaR

real    5m4.095s
user    7m27.847s
sys    1m26.448s
```

## [dirk/go-ipfs exp/add-perf](https://github.com/ipfs/go-ipfs/commit/06aa962e03e8086ab714a15f819946ae35d0b5bb)

```
$ time $GOPATH/bin/ipfs add -r --progress --offline --quieter ~/arch
27.07 TiB / 47.50 GiB [======================================================] 58362.03%
QmY5C4FXTDtmbyyKX1KixVMedJQ2nk9RgJMv6j9pDzrUaR

real    2m33.295s
user    6m54.554s
sys    1m44.620s
```

_note: the progress bar is currently a little broken 📈 _

**Comment by dirkmc** [comment]
Andrew also ran a test against the same Arch linux repo, where the repo was stored on an external hard drive. He found that there was no performance improvement of the new branch against the old branch for this use case:

### ipfs/go-ipfs master
```
time $GOPATH/bin/ipfs add -r --progress --offline --quieter /Volumes/5TB/arch/
47.50 GiB / 47.50 GiB [===============================================] 100.00%
QmY5C4FXTDtmbyyKX1KixVMedJQ2nk9RgJMv6j9pDzrUaR

real    11m2.681s
user    7m11.504s
sys    1m25.566s
```

### dirk/go-ipfs exp/add-perf
```
$ time $GOPATH/bin/ipfs add -r --progress --offline --quieter /Volumes/5TB/arch/
 2.90 TiB / 47.50 GiB [========================================================] 6255.61%5179323 milli-seconds add
674354 milli-seconds cumulative
QmY5C4FXTDtmbyyKX1KixVMedJQ2nk9RgJMv6j9pDzrUaR

real    11m15.050s
user    5m45.568s
sys    1m25.589s
```

This implies that there are probably two bottlenecks:
1. Chunking / IPLDing / CIDing (CPU bound)
2. Writing to DAGService (I/O bound)

I'd suggest we try adding a second queue to parallelize calls to DAGService. This is a little more complicated to implement as the calls to DAGService happen in several places (not just in go-ipfs)

**Comment by Stebalien** [comment]
> I'd suggest we try adding a second queue to parallelize calls to DAGService. This is a little more complicated to implement as the calls to DAGService happen in several places (not just in go-ipfs)

We already pipeline and batch writes to the dag service. Unfortunately, we flush the pipeline on every file, IIRC.

**Comment by Stebalien** [comment]
Ok, so @alanshaw and I have discussed this in https://github.com/ipfs/interface-js-ipfs-core/pull/509 and adding files out-of-order is technically supported and, really, there's no theoretical reason why it shouldn't be performant. I think the right path forward here is to fix/rewrite go-mfs.

**Comment by dirkmc** [comment]
@Stebalien it seems like if we allow out-of-order adds we'd need to keep the whole directory tree in a cache, whether we use MFS or not. Do you have an idea that will allow us to avoid that?

**Comment by Stebalien** [comment]
> @Stebalien it seems like if we allow out-of-order adds we'd need to keep the whole directory tree in a cache, whether we use MFS or not. Do you have an idea that will allow us to avoid that?


Explicitly open/close directories and eager-flush on close. In the best case, the overhead would be `O(depth)` memory (trivial). In the _worst_ case, the overhead would be a ton of extra hashing as we'd have to re-open previously closed (and hashed) directories, modify them, and then re-hash them.

**Comment by dirkmc** [comment]
I see so in that case we'd be optimizing for in-order adds, and doing some thrashing if adds are supplied out-of-order. That seems like a reasonable trade-off to me 👍

**Comment by Stebalien** [comment]
Yep. Ideally, we'd also make this configurable. That is, MFS could be configured to keep an LRU cache of open directories.

**Comment by Stebalien** [comment]
Had a good discussion with @achingbrain. Takeaways and questions:

* Does the periodic MFS flush stop the world? As in, is that causing us to do a bunch of sequential disk writes every so often, blocking everything else?
* When writing to an external HDD (WRT @andrew's tests), are we maxing out write throughput _anyways_. We can check this with tools like `iotop`. If we are than we may be barking up the wrong tree. @andrew, could you test this?
* What about adding a directory that's already _mostly_ added to IPFS? The internal "do we already have this block" checks should be saving us disk _writes_. Where are we bottlenecking here?

---

We also discussed the filestore a bit and the terrible UX. I believe a bunch of the bad UX comes from the fact that we're mixing pinning with the filestore.

Alternatively, we could _not_ do that. That is, we could say that all _pinned_ blocks must live in the datastore, and "filestore" blocks can only live in `MFS` (`ipfs files`) or in some special set of "synced" directories. That way overwriting or removing a file in the filestore wouldn't really _break_ anything. (note: fixing this would be a pretty large task).

**Comment by aschmahmann** [comment]
Did some analysis here and I'm having trouble reproducing a slow ipfs add.

**Definition:** A slow ipfs add is one where recursively adding a directory `D` takes more than 50% longer than directly copying the `D` (e.g. `cp -r, dd, robocopy, etc.`). In practice I've rarely seen anything more than 10-20%, but 50% should protect us from measurement variance.

**Test setup:** I've been running the tests on an 8TB HDD with the benchmark below, on Windows 10:

```
CrystalDiskMark 6.0.2 x64 (C) 2007-2018 hiyohiyo
                          Crystal Dew World : https://crystalmark.info/
-----------------------------------------------------------------------
* MB/s = 1,000,000 bytes/s [SATA/600 = 600,000,000 bytes/s]
* KB = 1000 bytes, KiB = 1024 bytes

   Sequential Read (Q= 32,T= 1) :   252.674 MB/s
  Sequential Write (Q= 32,T= 1) :   252.276 MB/s
  Random Read 4KiB (Q=  8,T= 8) :     1.329 MB/s [    324.5 IOPS]
 Random Write 4KiB (Q=  8,T= 8) :     1.219 MB/s [    297.6 IOPS]
  Random Read 4KiB (Q= 32,T= 1) :     1.287 MB/s [    314.2 IOPS]
 Random Write 4KiB (Q= 32,T= 1) :     1.142 MB/s [    278.8 IOPS]
  Random Read 4KiB (Q=  1,T= 1) :     0.756 MB/s [    184.6 IOPS]
 Random Write 4KiB (Q=  1,T= 1) :     1.055 MB/s [    257.6 IOPS]

  Test : 1024 MiB [E: 0.0% (0.3/7452.0 GiB)] (x3)  [Interval=5 sec]
  Date : 2019/10/07 15:20:04
    OS : Windows 10 Professional [10.0 Build 18362] (x64)
```

## Test 1: Add arch repo to go-ipfs 0.4.22 and also copy paste the repo

I used WSL to rsync the repo, but because of WSL weirdness the directory symlinks are annoying to work with so I only interacted with the pool directory (which holds 80% of the data as shown in https://github.com/ipfs/package-managers/issues/79).

Note: robocopy is generally considered a pretty fast Windows copy-paste utility, but if someone has a better alternative I'm open to it.

Folder Size: 38.6 GB 

Powershell:
```ps
ipfs init --profile=badgerds --empty-repo
ipfs pin rm QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn #removes only pin
ipfs repo gc

Measure-Command{ipfs22 add -r --offline --silent .\addtest\arch\arch\pool\};Measure-Command {robocopy /MIR /NFL /NDL .\addtest\arch\arch\pool\ .\testarch\} 
```

Results (minutes:seconds):
`ipfs add` : 08:48
`robocopy`: 08:33

## Test 2: Add extrapolated arch repo to go-ipfs 0.4.22 and also copy paste the repo

I tried running @dirkmc s tests above and got similar results to what he did on both the add and copy-paste. I then modified the tests to more precisely follow the distribution from https://github.com/ipfs/package-managers/issues/79. I unfortunately neglected to include the functionality allowing file sizes to be non-exponents of 2. I doubt that will really change the results, but I will post an update after I re-run the code.

Folder parameters:
Max directory width : 32
Number of files : 256k
Resulting directory size : 319GB

Powershell:
```ps
ipfs init --profile=badgerds --empty-repo
ipfs pin rm QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn #removes only pin
ipfs repo gc

Measure-Command{ipfs22 add -r --offline --silent .\addtest\256k\};Measure-Command {robocopy /MIR /NFL /NDL .\addtest\256k\ .\test256k\}
```

Results (hours:minutes:seconds):
`ipfs add` : 01:28:02
`robocopy`: 01:23:31

**Comment by Kubuxu** [comment]
@aschmahmann Windows is quite known for its slow filesystem. I would suggest trying on Linux.

**Comment by aschmahmann** [comment]
@Kubuxu while it's possible that Windows is not experiencing issues that OS X/Linux do either bc the file system is slow, or bc there's some bug in OS X/Linux not present in Windows, I have yet to see any evidence that this is a problem on OS X/Linux either.

Additionally, the arch test is basically the same as the one @dirkmc and @andrew ran and runs in a similar amount of time. So at least at the 40 GB range the speed differences don't appear to be super noticeable.

**Comment by dirkmc** [comment]
@aschmahmann could you try replicating this walk-through that Andrew wrote up: https://github.com/ipfs/package-managers/issues/18

**Comment by aschmahmann** [comment]
@dirkmc I can try, but it's going to be a little rough depending on how accurate I want the replication to be. I alluded to above I rsynced the Ubuntu repo (1.2 TB) in WSL which took quite a long time, but the WSL links won't resolve in Windows. I'm going to see if there's a way I can do one of:

A) Turn the WSL links to Windows links and use IPFS on Windows
B) Try to get WSL data into Linux and run the whole thing in Linux
C) Just use IPFS in WSL, but after doing some performance checks

**Comment by aschmahmann** [comment]
## Test 3: Add entire ubuntu repo to go-ipfs 0.4.22 and also copy paste the repo

I used deleted the WSL links and rsync'd as admin and got normal Windows links! (Windows normally requires admin to create symlinks, it doesn't seem like this shouldn't be necessary in Windows 10 developer mode but occurred nonetheless). So with my now Windows compatible WSL-rsync'd ubuntu repo...

Folder parameters:
Number of files : 958,560
Number of directories: 59,057
Resulting directory size : 1.24TB

Powershell:
```ps
ipfs init --profile=badgerds --empty-repo
ipfs pin rm QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn #removes only pin
ipfs repo gc

Measure-Command{ipfs22 add -r --offline --silent .\ubuntu\};Measure-Command {robocopy /MIR /NFL /NDL .\ubuntu\ .\testubuntu}
```

Results (hours:minutes:seconds):
`ipfs add` : 05:34:06
`robocopy`: (will test tomorrow, but expected to have similar performance since IPFS Add is running at the same ~60MB/s rate as it and robocopy did in the other tests)

Looks like `ipfs add` performance time is pretty comparable to copy paste. While it's possible we might be able to achieve better (e.g. since we're using a database and don't need many small files), this seems out of scope since as long as add performance is reasonable what we really need to work on is update performance (which @djdv is so kindly working on via his `ipfs mount` implementation work).

Something I noticed was that towards the very end of the test I came back to my machine and noticed that my RAM was maxed out (I have only 16GB of ram), I was hitting many hard faults, and disk read speeds were very low. I did some brief looking at the ram with [VMMap](https://docs.microsoft.com/en-us/sysinternals/downloads/vmmap) and it looks like a lot of it is from Badger (1.3 TB of mapped files, but also 10GB of ram allocated). It looks like we're probably using more ram than we should, but not necessarily by a huge margin. For those of you more experienced then me at looking through these ram allocation things I've attached a dump a took towards the end of the program running [ipfs22.txt](https://github.com/ipfs/go-ipfs/files/3770139/ipfs22.txt).

Given that this issue was filed because of [the ubuntu add experiment being slow](https://github.com/ipfs/package-managers/issues/18) and that I'm having trouble replicating it on reasonable hardware I recommend we pause this issue for now until a user comes back to us with a data set that they're having trouble importing.

Note, I haven't tested this on Linux or OS X yet, but as @Kubuxu mentioned Windows is supposed to be the slow one and it's doing just fine 😄. If you feel like testing on another OS is important here, please post away and we can discuss.

@Stebalien @dirkmc  thoughts?

**Comment by dirkmc** [comment]
It would be great if it's performing well enough that we can reprioritize other projects above this one, I'm glad you did this research.

It would be good to understand why the use case Andrew was looking at was so slow. It looks like he was doing this on Ubuntu, so maybe there's an issue on Linux that doesn't manifest on Windows? It's also possible that the hardware he was using wasn't up to the task, perhaps because it was severely memory limited or had a very slow disk.

**Comment by aschmahmann** [comment]
It looks from https://github.com/ipfs/package-managers/issues/18 like his machine had plenty of 32GB of ram (my initial though too). No idea what the drive performance was though, it was a cloud box but data wasn't in my face when I went to the website.

If I have time will try and run this on Ubuntu over the weekend (just got to get my machine set up). I'll probably use ntfs drivers instead of ext4 to avoid another 6hr file copy though.

**Comment by magik6k** [comment]
Note that NTFS on Linux has a way different performance characteristic than Ext4. From the datastore benchmarking dataset (done on c5d instance on aws, nvme ssd, not the greatest / cleanest dataset, but gives the idea) - https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ

Badger:
![](https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ/c5d-deb/ds-badger/add/prime-count-MBs-log.png)
![](https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ/c5d-deb/ds-badger/get/prime-count-nsop-log.png)
FlatFs:
![](https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ/c5d-deb/ds-flatfs/add/prime-count-MBs-log.png)
![](https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ/c5d-deb/ds-flatfs/get/prime-count-nsop-log.png)
LevelDB:
![](https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ/c5d-deb/ds-leveldb/add/prime-count-MBs-log.png)
![](https://ipfs.io/ipfs/QmPjUTbqYAsHfeuSsqZtPLZueH4r9oPnyj1kKckca9XANJ/c5d-deb/ds-leveldb/get/prime-count-nsop-log.png)

Also some notes:
* In me experience Windows is really slow when it comes to small IOs.
  * WSL being way slower that that
* If we are testing adding small files, those tests should happen on fast SSDs, which is where the problem is most visible, testing on HDDs with 1000x less IOPS than a decent SSD won't reveal any problems

**Comment by aschmahmann** [comment]
@magik6k 

> If we are testing adding small files, those tests should happen on fast SSDs, which is where the problem is most visible, testing on HDDs with 1000x less IOPS than a decent SSD won't reveal any problems

If the problem only exists on SSDs then according to @Stebalien it's fine to deprioritize this issue, since if we're really focusing on 1TB+ package mirrors they will likely be stored on HDDs (the issue that spun this off ipfs/package-managers#18 that also used HDDs).

Unless I'm misreading your first graph (add performance on badger) the results are confusing. It looks like NTFS is by far the fastest and for some reason ext4 performance is by far the worst. Is this accurate? If so then perhaps we're dealing with badger+ext4 issues.

**Comment by magik6k** [comment]
> It looks like NTFS is by far the fastest and for some reason ext4 performance is by far the worst

That's the correct observation, though it may indicate some issues with how ntfs implementation on Linux handles sync writes (given that it's faster than everything else too)

> If the problem only exists on SSDs then according to @Stebalien it's fine to deprioritize this issue, since if we're really focusing on 1TB+ package mirrors they will likely be stored on HDDs

That's fair, and I agree with this decision, but we should keep in mind that there are scenarios where we can still get better.

**Comment by aschmahmann** [comment]
## Test 1-ext4: Add arch repo to go-ipfs 0.4.22 and also copy paste the repo

I rsync'd the data from the NTFS partition to a blank ext4 partition on the same drive. Then tried `ipfs add` with Badger, and `cp -a`.

Powershell:
```ps
ipfs init --profile=badgerds --empty-repo
ipfs pin rm QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn #removes only pin
ipfs repo gc

Measure-Command {ipfs22 add -r --offline --silent ./arch/}
Measure-Command {cp -a ./arch/ ./testarch/}
```

Results (minutes:seconds):
`ipfs add` : 24:16
`cp -a`: 8:12

This seems to match with our expectations of ext4+badger being very slow and not with @dirkmc and @andrew 's results. @dirkmc any idea if the partitions were ext4 formatted and if they were running on HDDs or SSDs?

**Comment by dirkmc** [comment]
Agreed, these are really useful findings, thanks Adin.

My tests were all done on my mac with an SSD and plenty of RAM. I'm not sure about the characteristics of the machine Andrew was using originally.

**Comment by magik6k** [comment]
> were the NTFS results collected on Linux or Windows?

That was on one of the Debian versions AWS provided at the time, not sure exactly which one

**Comment by aschmahmann** [comment]
@dirkmc @magik6k thanks for the info.

I ran more tests and Ubuntu 19.10+badger+ext4 performs much much better if you turn off `syncwrites` in badger (doable as an IPFS config option), and performance continues to increase if you turn off ext4 journaling (I updated results in the post above).

## Test 3-ext4: Add entire ubuntu repo to go-ipfs 0.4.22

I then went and ran the ubuntu repo test after turning off sync writes and ext4 journaling. For those interested in turning off journaling I followed the slightly wrong instructions at https://foxutech.com/how-to-disable-enable-journaling/. Instructions for posterity:

```
MyExt4MountPt = /media/me/myextpartiton
MyExt4Partition = /dev/sda3
#umount MyExt4MountPt 
#tune2fs -O ^has_journal MyExt4Partition
#e2fsck –f MyExt4Partition #optional health check, bc why not

reboot

#dmesg | grep EXT4 #this should show the MyExt4Partition has journaling disabled
```

Powershell:
```ps
ipfs init --profile=badgerds --empty-repo
ipfs pin rm QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn #removes only pin
ipfs repo gc
#manually edited IPFS config file to use "syncWrites" : false, could probably have used the cli

Measure-Command{ipfs22 add -r --offline --silent .\ubuntu\}
```

Results (hours:minutes:seconds):
`ipfs add` : 05:51:34

This super snazzy result (plus some profiling I did with pprof - thanks @dirkmc for the easy setup) indicate that the problem is that linux+ext4+badger is really bad with `syncwrite` turned on. There are a number of ways we can deal with this going forward.

Two main camps going forward are probably:
1) Figure out why this is happening, is it linux or ext4 (or the combination) that's causing the badger problem. Do they have a bug or is it system to how linux and/or ext4 work?
2) Who care why it happens, let's just make our transactions better so that we effectively emulate `syncwrite = false` when doing `ipfs add`.

My plan is to go for option 2 as it is the fastest way to get results. May also be worth checking in with the badger folks to see if they have a bug. Perhaps it's related to https://github.com/dgraph-io/badger/issues/1084 ?

**Comment by dirkmc** [comment]
Amazing ❤️

**Comment by Stebalien** [comment]
# Improve Add Performance W/ Async Datastores

* [ ] Add async write support to datastores - https://github.com/ipfs/go-datastore/issues/137
* [ ] Add `Sync` calls where appropriate (e.g., when pinning).
* [ ] Turn syncwrites off on badger in go-ipfs by default.
* [ ] Switch to badger by default - https://github.com/ipfs/go-ipfs/issues/4279.

**Comment by momack2** [comment]
@aschmahmann - I created a placeholder graph [here](https://docs.google.com/spreadsheets/d/1LK1qj95WnM6JJEQa-4OTWn5wpZtRRU4HxHp76n4YUAI/edit#gid=1347953717) for us to organize this information. Can you help add in the right data for 0.5 vs 0.4.23 using badger vs flat fs on linux?

**Comment by phillmac** [comment]
Just my 2¢, Is it possible to include benchmarks for both ext4, btrfs and optionally ntfs underlying filesystems? I'm in the position of deploying some VMs dedicated purely to ipfs related tasks, and having a recommended file system type for maximum performance tuning seems appropriate and useful imho.

**Comment by Stebalien** [comment]
Unfortunately, running these benchmarks is a manual process so we're not going to be able to re-run them on all filesystems. However, if you want to try a multi-filesystem benchmark, I'd be interested in the results.

(in general, I'd recommend against NTFS).

**Comment by alexlatif** [comment]
Has this been abandoned?

**Comment by lidel** [comment]
Closing this as duplicate of https://github.com/ipfs/kubo/issues/10383
